File: 02 - Breadth First Search/M605.SequenceReconstruction.py
class Solution:
    """
    @param org: a permutation of the integers from 1 to n
    @param seqs: a list of sequences
    @return: true if it can be reconstructed only one or false
    """
    def sequenceReconstruction(self, org, seqs):
        # write your code here
        graph = self.build_graph(seqs)
        result = self.sort_topological(graph)

        return result == org

    def build_graph(self, seqs):
        # initialize a graph
        graph = {}

        for seq in seqs:
            for i in seq:
                if i not in graph:
                    graph[i] = set()

        for seq in seqs:
            for i in range(1, len(seq)):
                graph[seq[i - 1]].add(seq[i])

        return graph

    def sort_topological(self, graph):
        indegrees, queue = self.get_indegrees(graph), []

        for node in graph:
            if not indegrees[node]:
                queue.append(node)

        result = []
        while queue:
            # need only one sequence
            if len(queue) > 1:
                return None

            node = queue.pop()
            result.append(node)
            for neighbor in graph[node]:
                indegrees[neighbor] -= 1
                if not indegrees[neighbor]:
                    queue.append(neighbor)

        if len(result) == len(graph):
            return result

        return None

    def get_indegrees(self, graph):
        indegrees = {node: 0 for node in graph}

        for node in graph:
            for neighbor in graph[node]:
                indegrees[neighbor] += 1

        return indegrees

File: 02 - Breadth First Search/M605.SequenceReconstruction.java
public class Solution {
    /**
     * @param org: a permutation of the integers from 1 to n
     * @param seqs: a list of sequences
     * @return: true if it can be reconstructed only one or false
     */
    public boolean sequenceReconstruction(int[] org, int[][] seqs) {
        // write your code here
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        Map<Integer, Integer> indegree = new HashMap<>();

        for (int num: org) {
            graph.put(num, new HashSet<>());
            indegree.put(num, 0);
        }

        int n = org.length, m = 0;
        for (int[] seq: seqs) {
            m += seq.length;
            if (seq.length >= 1 && (seq[0] <= 0 || seq[0] > n)) return false;

            for (int i = 1; i < seq.length; i++) {
                if (seq[i] <= 0 || seq[i] > n) return false;

                if (graph.get(seq[i - 1]).add(seq[i])) {
                    indegree.put(seq[i], indegree.get(seq[i]) + 1);
                }
            }
        }

        if (m < n) return false;

        Queue<Integer> q = new LinkedList<>();
        for (int key: indegree.keySet()) {
            if (indegree.get(key) == 0) q.offer(key);
        }

        int count = 0;
        while (q.size() == 1) {
            int num = q.poll();
            for (int next: graph.get(num)) {
                indegree.put(next, indegree.get(next) - 1);
                if (indegree.get(next) == 0) q.offer(next);
            }

            if (num != org[count++]) return false;
        }

        return count == n;
    }
}
